diff -ipuwr -x config.log -x config.status -x config.cache -x '*.Po' -x '*.pot' -x '*.bak' -x '*.orig' -x '*.rej' -x '*.new' -x '*~' -x '*.o' -x '*.exe' -x patches bzip2-1.0.2-orig/Makefile bzip2-1.0.2/Makefile
--- bzip2-1.0.2-orig/Makefile	2002-01-25 23:34:54.000000000 +0100
+++ bzip2-1.0.2/Makefile	2004-06-09 22:29:48.796875000 +0200
@@ -1,3 +1,6 @@
+VERSION = 1.02
+MAJOR=$(shell echo $(VERSION) | sed -e "s/\\..*$$//")
+MINOR=$(shell echo $(VERSION) | sed -e "s/^[^\\.]*\\.0*\\([^.]*\\).*$$/\\1/")
 
 SHELL=/bin/sh
 
@@ -5,14 +8,20 @@ SHELL=/bin/sh
 CC=gcc
 AR=ar
 RANLIB=ranlib
-LDFLAGS=
+LDFLAGS= -Wl,-s	-Wl,--major-image-version=$(MAJOR) -Wl,--minor-image-version=$(MINOR)
+EXE_EXT=.exe
+DLL_EXT=.dll
+LN=cp -fp
 
 # Suitably paranoid flags to avoid bugs in gcc-2.7
-BIGFILES=-D_FILE_OFFSET_BITS=64
-CFLAGS=-Wall -Winline -O2 -fomit-frame-pointer -fno-strength-reduce $(BIGFILES)
+BIGFILES = -D_LARGEFILE_SOURCE=1 -D_LARGEFILE64_SOURCE=1 -D_FILE_OFFSET_BITS=64
+CFLAGS = -D__GW32__ -Wall -Winline -O3 -mms-bitfields  -fno-exceptions \
+	-fomit-frame-pointer -march=i386 -mcpu=i686 -fno-strength-reduce $(BIGFILES) \
+	-idirafter h:/glibc/include
+LIBS = -lstdbin -lbinmode -lgw32c -lole32 -luuid -liberty
 
 # Where you want it installed when you do 'make install'
-PREFIX=/usr
+PREFIX = c:/progra~1/Bzip2
 
 
 OBJS= blocksort.o  \
@@ -23,13 +32,21 @@ OBJS= blocksort.o  \
       decompress.o \
       bzlib.o
 
-all: libbz2.a bzip2 bzip2recover test
+DLLOBJS = $(OBJS) bzip-dllversion.o
 
-bzip2: libbz2.a bzip2.o
-	$(CC) $(CFLAGS) $(LDFLAGS) -o bzip2 bzip2.o -L. -lbz2
+# Rule to make compiled resource (Windows)
+%-rc.o: %.rc
+	windres -i $< -o $@
 
-bzip2recover: bzip2recover.o
-	$(CC) $(CFLAGS) $(LDFLAGS) -o bzip2recover bzip2recover.o
+all: libbz2.a bzip2$(DLL_EXT) bzip2 bzip2recover test
+
+bzip2: libbz2.a bzip2.o bzip2-rc.o
+	$(CC) $(CFLAGS) $(LDFLAGS) -o bzip2 bzip2.o bzip2-rc.o \
+		./libbz2.a $(LIBS)
+
+bzip2recover: bzip2recover.o bzip2recover-rc.o
+	$(CC) $(CFLAGS) $(LDFLAGS) -o bzip2recover bzip2recover.o \
+		bzip2recover-rc.o $(LIBS)
 
 libbz2.a: $(OBJS)
 	rm -f libbz2.a
@@ -40,6 +57,9 @@ libbz2.a: $(OBJS)
 		$(RANLIB) libbz2.a ; \
 	fi
 
+bzip2$(DLL_EXT): $(DLLOBJS)
+	o2dll -o $@ -l libbz2 $(DLLOBJS)
+
 check: test
 test: bzip2
 	@cat words1
@@ -63,31 +83,41 @@ install: bzip2 bzip2recover
 	if ( test ! -d $(PREFIX)/man ) ; then mkdir -p $(PREFIX)/man ; fi
 	if ( test ! -d $(PREFIX)/man/man1 ) ; then mkdir -p $(PREFIX)/man/man1 ; fi
 	if ( test ! -d $(PREFIX)/include ) ; then mkdir -p $(PREFIX)/include ; fi
-	cp -f bzip2 $(PREFIX)/bin/bzip2
-	cp -f bzip2 $(PREFIX)/bin/bunzip2
-	cp -f bzip2 $(PREFIX)/bin/bzcat
-	cp -f bzip2recover $(PREFIX)/bin/bzip2recover
-	chmod a+x $(PREFIX)/bin/bzip2
-	chmod a+x $(PREFIX)/bin/bunzip2
-	chmod a+x $(PREFIX)/bin/bzcat
-	chmod a+x $(PREFIX)/bin/bzip2recover
-	cp -f bzip2.1 $(PREFIX)/man/man1
+	cp -fp bzip2$(EXE_EXT) $(PREFIX)/bin/bzip2$(EXE_EXT)
+	cp -fp bzip2$(DLL_EXT) $(PREFIX)/bin/bzip2$(DLL_EXT)
+	cp -fp bzip2$(EXE_EXT) $(PREFIX)/bin/bunzip2$(EXE_EXT)
+	cp -fp bzip2$(EXE_EXT) $(PREFIX)/bin/bzcat$(EXE_EXT)
+	cp -fp bzip2recover$(EXE_EXT) $(PREFIX)/bin/bzip2recover$(EXE_EXT)
+	chmod a+x $(PREFIX)/bin/bzip2$(EXE_EXT)
+	chmod a+x $(PREFIX)/bin/bzip2$(DLL_EXT)
+	chmod a+x $(PREFIX)/bin/bunzip2$(EXE_EXT)
+	chmod a+x $(PREFIX)/bin/bzcat$(EXE_EXT)
+	chmod a+x $(PREFIX)/bin/bzip2recover$(EXE_EXT)
+	cp -fp bzip2.1 $(PREFIX)/man/man1
 	chmod a+r $(PREFIX)/man/man1/bzip2.1
-	cp -f bzlib.h $(PREFIX)/include
+	cp -fp bzlib.h $(PREFIX)/include
 	chmod a+r $(PREFIX)/include/bzlib.h
-	cp -f libbz2.a $(PREFIX)/lib
+	cp -fp libbz2.a $(PREFIX)/lib
+	cp -fp libbz2$(DLL_EXT).a $(PREFIX)/lib
+	cp -fp libbz2.lib $(PREFIX)/lib
+	cp -fp libbz2-bcc.lib $(PREFIX)/lib
+	cp -fp bzip2.def $(PREFIX)/lib
 	chmod a+r $(PREFIX)/lib/libbz2.a
-	cp -f bzgrep $(PREFIX)/bin/bzgrep
-	ln $(PREFIX)/bin/bzgrep $(PREFIX)/bin/bzegrep
-	ln $(PREFIX)/bin/bzgrep $(PREFIX)/bin/bzfgrep
+	chmod a+r $(PREFIX)/lib/libbz2$(DLL_EXT).a
+	chmod a+r $(PREFIX)/lib/libbz2.lib
+	chmod a+r $(PREFIX)/lib/libbz2-bcc.lib
+	chmod a+r $(PREFIX)/lib/bzip2.def
+	cp -fp bzgrep $(PREFIX)/bin/bzgrep
+	$(LN) $(PREFIX)/bin/bzgrep $(PREFIX)/bin/bzegrep
+	$(LN) $(PREFIX)/bin/bzgrep $(PREFIX)/bin/bzfgrep
 	chmod a+x $(PREFIX)/bin/bzgrep
-	cp -f bzmore $(PREFIX)/bin/bzmore
-	ln $(PREFIX)/bin/bzmore $(PREFIX)/bin/bzless
+	cp -fp bzmore $(PREFIX)/bin/bzmore
+	$(LN) $(PREFIX)/bin/bzmore $(PREFIX)/bin/bzless
 	chmod a+x $(PREFIX)/bin/bzmore
-	cp -f bzdiff $(PREFIX)/bin/bzdiff
-	ln $(PREFIX)/bin/bzdiff $(PREFIX)/bin/bzcmp
+	cp -fp bzdiff $(PREFIX)/bin/bzdiff
+	$(LN) $(PREFIX)/bin/bzdiff $(PREFIX)/bin/bzcmp
 	chmod a+x $(PREFIX)/bin/bzdiff
-	cp -f bzgrep.1 bzmore.1 bzdiff.1 $(PREFIX)/man/man1
+	cp -fp bzgrep.1 bzmore.1 bzdiff.1 $(PREFIX)/man/man1
 	chmod a+r $(PREFIX)/man/man1/bzgrep.1
 	chmod a+r $(PREFIX)/man/man1/bzmore.1
 	chmod a+r $(PREFIX)/man/man1/bzdiff.1
@@ -98,9 +128,10 @@ install: bzip2 bzip2recover
 
 distclean: clean
 clean: 
-	rm -f *.o libbz2.a bzip2 bzip2recover \
+	rm -f *.o libbz2*.a bzip2$(EXE_EXT) bzip2recover$(EXE_EXT) bzip2$(DLL_EXT) \
 	sample1.rb2 sample2.rb2 sample3.rb2 \
-	sample1.tst sample2.tst sample3.tst
+	sample1.tst sample2.tst sample3.tst	 \
+	libbz2*.lib 
 
 blocksort.o: blocksort.c
 	@cat words0
@@ -125,7 +156,7 @@ bzip2recover.o: bzip2recover.c
 DISTNAME=bzip2-1.0.2
 tarfile:
 	rm -f $(DISTNAME)
-	ln -sf . $(DISTNAME)
+	$(LN) . $(DISTNAME)
 	tar cvf $(DISTNAME).tar \
 	   $(DISTNAME)/blocksort.c \
 	   $(DISTNAME)/huffman.c \
@@ -161,7 +192,7 @@ tarfile:
 	   $(DISTNAME)/README \
 	   $(DISTNAME)/README.COMPILATION.PROBLEMS \
 	   $(DISTNAME)/CHANGES \
-	   $(DISTNAME)/libbz2.def \
+	   $(DISTNAME)/bzip2.def \
 	   $(DISTNAME)/libbz2.dsp \
 	   $(DISTNAME)/dlltest.dsp \
 	   $(DISTNAME)/makefile.msc \
diff -ipuwr -x config.log -x config.status -x config.cache -x '*.Po' -x '*.pot' -x '*.bak' -x '*.orig' -x '*.rej' -x '*.new' -x '*~' -x '*.o' -x '*.exe' -x patches bzip2-1.0.2-orig/bzip2.c bzip2-1.0.2/bzip2.c
--- bzip2-1.0.2-orig/bzip2.c	2002-01-25 01:38:36.000000000 +0100
+++ bzip2-1.0.2/bzip2.c	2004-03-11 15:15:14.953125000 +0100
@@ -127,8 +127,10 @@
 #define BZ_LCCWIN32  0
 
 #if defined(_WIN32) && !defined(__CYGWIN__)
+# ifndef __GW32__
 #undef  BZ_LCCWIN32
 #define BZ_LCCWIN32 1
+# endif
 #undef  BZ_UNIX
 #define BZ_UNIX 0
 #endif
@@ -221,6 +223,7 @@
 #   include <io.h>
 #   include <fcntl.h>
 #   include <sys\stat.h>
+#   include <sys\utime.h>
 
 #   define NORETURN       /**/
 #   define PATH_SEP       '\\'
@@ -244,6 +247,37 @@
 
 #endif /* BZ_LCCWIN32 */
 
+#if __GW32__
+#   include <fcntl.h>
+#   include <sys/types.h>
+#   include <utime.h>
+#   include <unistd.h>
+#   include <sys/stat.h>
+#   include <sys/times.h>
+#   include <io.h>
+
+#   define NORETURN       /**/
+#   define PATH_SEP       '\\'
+#   define MY_LSTAT       lstat
+#   define MY_STAT        stat
+#   define MY_S_ISREG(x)  ((x) & _S_IFREG)
+#   define MY_S_ISDIR(x)  ((x) & _S_IFDIR)
+
+#   define APPEND_FLAG(root, name) \
+      root=snocString((root), (name))
+
+#   define APPEND_FILESPEC(root, name)                \
+      root = snocString ((root), (name))
+
+#   define SET_BINARY_MODE(fd)                        \
+      do {                                            \
+         int retVal = setmode ( fileno ( fd ),        \
+                                O_BINARY );           \
+         ERROR_IF_MINUS_ONE ( retVal );               \
+      } while ( 0 )
+
+#endif /* __GW32__ */
+
 
 /*---------------------------------------------*/
 /*--
@@ -1107,7 +1141,7 @@ Int32 countHardLinks ( Char* name )
    robustly to arbitrary Unix-like platforms (or even works robustly
    on this one, RedHat 7.2) is unknown to me.  Nevertheless ...  
 */
-#if BZ_UNIX
+#if defined(BZ_UNIX) || defined (_WIN32)
 static 
 struct MY_STAT fileMetaInfo;
 #endif
@@ -1115,7 +1149,7 @@ struct MY_STAT fileMetaInfo;
 static 
 void saveInputFileMetaInfo ( Char *srcName )
 {
-#  if BZ_UNIX
+#if defined(BZ_UNIX) || defined (_WIN32)
    IntNative retVal;
    /* Note use of stat here, not lstat. */
    retVal = MY_STAT( srcName, &fileMetaInfo );
@@ -1127,7 +1161,7 @@ void saveInputFileMetaInfo ( Char *srcNa
 static 
 void applySavedMetaInfoToOutputFile ( Char *dstName )
 {
-#  if BZ_UNIX
+#if defined(BZ_UNIX) || defined (_WIN32)
    IntNative      retVal;
    struct utimbuf uTimBuf;
 
@@ -1140,11 +1174,13 @@ void applySavedMetaInfoToOutputFile ( Ch
    retVal = utime ( dstName, &uTimBuf );
    ERROR_IF_NOT_ZERO ( retVal );
 
+#  if BZ_UNIX
    retVal = chown ( dstName, fileMetaInfo.st_uid, fileMetaInfo.st_gid );
    /* chown() will in many cases return with EPERM, which can
       be safely ignored.
    */
 #  endif
+#  endif
 }
 
 
